### PR TITLE
Reimplement abi encode/decode using stream/sinks 

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

File: src/eth/mod.rs [hunk 1]
```diff
@@ -1,25 +1,53 @@
...
+
+/// Error for decoding rust types from stream
+#[derive(Debug)]
+pub enum Error {
+	/// Invalid bool for provided input
+	InvalidBool,
+	/// Invalid u32 for provided input
+	InvalidU32,
+	/// Invalid u32 for provided input
+	InvalidU64,
```
Comment (pepyakin) [inline src/eth/mod.rs]: comments says that it is u32

File: src/eth/common.rs [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+use lib::*;
+use super::{util, Stream, AbiType, Sink, Error};
+use parity_hash::Address;
+use bigint::U256;
+
+impl AbiType for u32 {
+	fn decode(stream: &mut Stream) -> Result<Self, Error> {
+		let previous_position = stream.advance(32);
+		if stream.position() > stream.payload().len() {
+			return Err(Error::UnexpectedEof);
```
Comment (pepyakin) [inline src/eth/common.rs]: I'm wondering if these checks could be moved into `advance`?
So it would be 
```rust
stream.advance(32)?;
```

Comment (NikVolf) [reply src/eth/common.rs]: Indeed

File: src/eth/common.rs [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	fn encode(self, sink: &mut Sink) {
+		let tail = sink.preamble_mut().len();
+		sink.preamble_mut().resize(tail + 32, 0);
+		sink.preamble_mut()[tail+12..tail+32].copy_from_slice(self.as_ref());
+	}
+
+	fn is_fixed() -> bool { true }
+}
+
+impl<T: AbiType> AbiType for Vec<T> {
```
Comment (pepyakin) [inline src/eth/common.rs]: Can we get away with &[T] ?

Comment (pepyakin) [reply src/eth/common.rs]: Ah, nevermind, we can't encode it.

**Pull request merged**